overlord/exportstate/special.go

```diff
@@ -0,0 +1,172 @@
+	"github.com/snapcore/snapd/release"
+	"github.com/snapcore/snapd/snap"
+)
+
+// special.go holds special cases when the export manifest is not stored in snap.yaml
+
+// exportSetSnapdTools returns an export set describing snapd tools.
+//
+// basePath is the path of the "$libexecdir/snapd" directory. This directory
+// differs depending on who is providing the tools. Host with variable
+// $libexecdir or snaps with a fixed one.
+func exportSetSnapdTools(basePath string) ExportSet {
+	// tools contains the list of snapd tools to export.
+	tools := []string{
+		"etelpmoc.sh",         // used by tab completion logic
+		"info",                // used by re-execution logic
+		"snap-confine",        // sometimes used in special cases
+		"snap-discard-ns",     // used by snap-confine inside the per-snap mount namespace
+		"snap-exec",           // used by snap-confine inside the per-snap mount namespace
+		"snap-gdb-shim",       // used by snap run --gdb
+		"snap-gdbserver-shim", // used by snap run --experimental-gdbserver
+		"snap-update-ns",      // used by snap-confine inside the per-snap mount namespace
+		"snapctl",             // used by snaps to talk to snapd
+	}
+	set := ExportSet{
+		Name:           "tools",
+		ConsumerIsHost: false, // Those tools are for other snaps to use.
+		Exports:        make(map[string]ExportedFile, len(tools)),
+	}
+	for _, tool := range tools {
+		set.Exports[tool] = ExportedFile{
+			Name:       tool,
+			SourcePath: filepath.Join(basePath, tool),
+		}
+	}
+	return set
+}
+
+// manifestForCoreSystem returns the manifest of the host as seen on ubuntu-core systems.
+func manifestForCoreSystem() *Manifest {
+	return &Manifest{
+		ExportedForSnapdAsVersion: "host", // Exception from the rule
+		SourceIsHost:              true,
+		// There are no export sets here, snapd snap is going to provide the tools.
+	}
+}
+
+// manifestForClassicSystem returns the manifest of the host as seen on classic systems.
+func manifestForClassicSystem() *Manifest {
+	tools := exportSetSnapdTools(dirs.DistroLibExecDir)
+	return &Manifest{
+		ExportedForSnapdAsVersion: "host", // Exception from the rule
+		SourceIsHost:              true,
+		Sets:                      map[string]ExportSet{tools.Name: tools},
+	}
+}
+
+// manifestForSnapdSnap returns the manifest of the snapd snap.
+func manifestForSnapdSnap(info *snap.Info) *Manifest {
+	tools := exportSetSnapdTools("/usr/lib/snapd")
+	return &Manifest{
+		SnapInstanceName: info.InstanceName(),
+		SnapRevision:     info.Revision,
+		Sets:             map[string]ExportSet{tools.Name: tools},
+	}
+}
+
+// manifestForCoreSnap returns the manifest of the core snap.
+func manifestForCoreSnap(info *snap.Info) *Manifest {
+	tools := exportSetSnapdTools("/usr/lib/snapd")
+	return &Manifest{
+		SnapInstanceName:          info.InstanceName(),
+		SnapRevision:              info.Revision,
+		ExportedForSnapdAsVersion: fmt.Sprintf("core_%s", info.Revision), // Exception from the rule
```

Comment:
it's still not clear to me why we do this translation and don't let this export to be named after the `core` snap itself, instead of exposing it as `snapd`.

Reply:
I eventually got it :-)